crates/factor-variables/src/provider_type.rs

```diff
@@ -0,0 +1,49 @@
+use std::{collections::HashMap, sync::Arc};
+
+use serde::{de::DeserializeOwned, Deserialize};
+use spin_expressions::{async_trait::async_trait, Key, Provider};
+use spin_factors::anyhow;
+
+pub trait VariablesProviderType: 'static {
+    const TYPE: &'static str;
+
+    type RuntimeConfig: DeserializeOwned;
+    type Provider: Provider;
+
+    fn make_provider(&self, runtime_config: Self::RuntimeConfig) -> anyhow::Result<Self::Provider>;
+}
+
+pub(crate) type ProviderMaker = Box<dyn Fn(toml::Table) -> anyhow::Result<Box<dyn Provider>>>;
+
+pub(crate) fn provider_maker<T: VariablesProviderType>(provider_type: T) -> ProviderMaker {
+    Box::new(move |table| {
+        let runtime_config: T::RuntimeConfig = table.try_into()?;
+        let provider = provider_type.make_provider(runtime_config)?;
+        Ok(Box::new(provider))
+    })
+}
+
+pub struct StaticVariables;
+
+impl VariablesProviderType for StaticVariables {
+    const TYPE: &'static str = "static";
+
+    type RuntimeConfig = StaticVariablesProvider;
+    type Provider = StaticVariablesProvider;
```

Comment:
I don't think you yet have an example where these two types differ. It would be nice to see that to make sure that the complexity of having two different types here is worth it (i.e., why not require the struct that represents the config also plays the role of the provider?)

Reply:
It will definitely be different for the vault provider.